spec/virtual_total_spec.rb

```diff
@@ -435,37 +435,144 @@ def model_with_children(count)
+          end.to_not make_database_queries
+        end
+
+        it "calculates avg from attribute (and preloaded association)" do
+          authors
+          query = Author.includes(:recently_published_books).select(:id, :avg_recently_published_books_rating).order(:id).load
+          expect do
+            expect(query.map(&:avg_recently_published_books_rating)).to eq([3, 5, 0, 0])
+          end.to_not make_database_queries
+        end
       end
 
-      it "calculates sum from attribute" do
-        authors
-        query = Author.select(:id, :sum_recently_published_books_rating).order(:id).load
-        expect do
-          expect(query.map(&:sum_recently_published_books_rating)).to eq([6, 5, 0, 0])
-        end.to make_database_queries(:count => 2)
+      describe ":maximum" do
+        # NOTE: rails converts the nil to a 0
+        it "calculates max with one off query" do
+          authors
+
+          expect do
+            expect(authors.map(&:highest_recently_published_books_rating)).to eq([4, 5, 0, 0])
+          end.to make_database_queries(:count => 4)
+        end
+
+        it "calculates max from preloaded association" do
+          authors.each { |a| a.recently_published_books.load }
+
+          expect do
+            expect(authors.map(&:highest_recently_published_books_rating)).to eq([4, 5, 0, 0])
+          end.to_not make_database_queries
+        end
+
+        it "calculates max from attribute" do
+          authors
+          query = Author.select(:id, :highest_recently_published_books_rating).order(:id).load
+          expect do
+            expect(query.map(&:highest_recently_published_books_rating)).to eq([4, 5, 0, 0])
+          end.to_not make_database_queries
+        end
+
+        it "calculates max from attribute (and preloaded association)" do
+          authors
+          query = Author.includes(:recently_published_books).select(:id, :highest_recently_published_books_rating).order(:id).load
+          expect do
+            expect(query.map(&:highest_recently_published_books_rating)).to eq([4, 5, 0, 0])
+          end.to_not make_database_queries
+        end
       end
 
-      it "calculates sum from attribute (and preloaded association)" do
-        authors
-        query = Author.includes(:recently_published_books).select(:id, :sum_recently_published_books_rating).order(:id).load
-        expect do
-          expect(query.map(&:sum_recently_published_books_rating)).to eq([6, 5, nil, 0])
-        end.to_not make_database_queries
+      describe ":minimum" do
+        # NOTE: rails converts the nil to a 0
+        it "calculates min with one off query" do
+          authors
+
+          expect do
+            expect(authors.map(&:lowest_recently_published_books_rating)).to eq([2, 5, 0, 0])
+          end.to make_database_queries(:count => 4)
+        end
+
+        it "calculates min from preloaded association" do
+          authors.each { |a| a.recently_published_books.load }
```

Comment:
we have the concept of a relation, (`recently_published_books`) and a attribute (`lowest_recently_published_books_rating`) and an expected value (`[2, 5, 0, 0]`)
from there I think we can introduce an example.

I guess the relation is authors, and the sub relation is recently published book ratings, ... naming is hard

Just for this page for now.